### ISSUE TITLE
Errors in check of repo converted from attic

### ISSUE THREAD

**Issue by cyril42e** [borgbackup/borg#1837]
Hi, I made a "borg check" on my compressed unencrypted repo stored on an USB external hard drive because of #1138 (with borg 1.0.8), and though repository analysis shows no error, during archives analysis I am getting plenty of errors such as :
`Did not get expected metadata dict when unpacking item metadata [chunk: 000026_a40edec797ea5a6c57452c3c5400b2566fffd75d86699fa6a144d207946237de]`
(with tens of errors for the same chunk, and thousands of chunks). Orphaned objects have been found as well.

I verified each archive independently, and it appears that only the older archives that were created with attic show this problem, not the archives that were created with borg after the conversion. The complete history of my repo is the following :
- September 2014 : created repo with attic 0.13
- September 2014 - October 2014 : created several archives with attic 0.13
- August 2015 : converted the repo with borg 0.23.0
- August 2015 - December 2015 : created several archives with borg 0.23.0 (and maybe other versions <= 0.29.0)

I ran "attic check" (with attic 0.16) on the original attic repository (which I kept), and it did not find any problem.

I focused on the latest archive created with attic (also the latest that has the problem). I attached the complete output of "borg check -v --archives-only" ([check.log.gz](https://github.com/borgbackup/borg/files/587393/check.log.gz)), and the result of "borg debug-dump-archive-items" ([items.log.gz](https://github.com/borgbackup/borg/files/587394/items.log.gz)). There are 4237 chunks, with 6 groups of duplicated chunks, each group with 2 or 3 duplicated chunks :
`
% ls | sort -k 1.8,1.24 | uniq -D -s 7
000752_06e338c1e28099ca2ad07ec728ba265f842db9754d21a94eb5f1e96bc320c517.items
000784_06e338c1e28099ca2ad07ec728ba265f842db9754d21a94eb5f1e96bc320c517.items
000444_1a5a6c364d47290cb2c776542ff3b8cd8158f31867eb46f2a396ea96ef2090d6.items
001098_1a5a6c364d47290cb2c776542ff3b8cd8158f31867eb46f2a396ea96ef2090d6.items
001137_1a5a6c364d47290cb2c776542ff3b8cd8158f31867eb46f2a396ea96ef2090d6.items
000761_2eabd1d3873fd061e0a24614e4d3ac9775421caa2ab96a37dfac1e3d69d7e454.items
000793_2eabd1d3873fd061e0a24614e4d3ac9775421caa2ab96a37dfac1e3d69d7e454.items
000750_4909761a9103abb057d075b9a4b4cf1c25971cb107feac064ca1362e4ff786a8.items
000782_4909761a9103abb057d075b9a4b4cf1c25971cb107feac064ca1362e4ff786a8.items
000751_fb370bb6b3edaa9e6c79fd75a7fca949041e655e4e8a6f39c17d6c6d727c9410.items
000783_fb370bb6b3edaa9e6c79fd75a7fca949041e655e4e8a6f39c17d6c6d727c9410.items
000450_fbd07dcf5e9e67fc579446d4e66b010afd7398a0384779afe8353935323541eb.items
001100_fbd07dcf5e9e67fc579446d4e66b010afd7398a0384779afe8353935323541eb.items
001139_fbd07dcf5e9e67fc579446d4e66b010afd7398a0384779afe8353935323541eb.items
`

Two of these groups are less than 4kB (2.4kB and 3.9kB).

Any idea of what could have happened ?

**Comment by enkore** [comment]
Is this the first check you ran, or did it previously report no errors, but now started to report these errors?

**Comment by cyril42e** [comment]
I probably ran a check after the conversion, but I don't remember for sure. I did not run another check afterwards, until today.

**Comment by ThomasWaldmann** [comment]
Hmm, a bit unusual for USB, but is this maybe same bug we fixed in 1.0.4? (see big warning at beginning of changes.rst)

**Comment by enkore** [comment]
This looks like we broke valid item detection in one of the more recent versions, because the item/chunk count is quite plausible and it was changed recently.

Can you take a look at one of these .items files it gave you, whether there is some private data in there (like file names /paths, xattrs, these are plain-text). If you could find one that you're comfortable sharing with us it would be very helpful. If you don't want to upload it publicly you can send it via mail to Thomas or me (emails in the AUTHORS file in the repo).

**Comment by cyril42e** [comment]
Sent some items by email. I forgot to mention that the repo is compressed (the archives are created with parameters "--compression zlib,6 --chunker-params 10,23,16,4095").
Let me know if there is anything else I can do or try!

**Comment by enkore** [comment]
Thanks a lot for the data, it would've been quite difficult to find the issue without it! :)

After some hex-pickaxing my eye caught onto 'acl'. It seems that early attic releases up to and including 0.13 had a bug that always, unconditionally stored the value None in the 'acl' key. This was fixed in jborg/attic@8f4e0f7506eee280d5489412aa5dde955dd6fed6 ::

```py
    def stat_attrs(self, st, path):
        ...
        item[b'acl'] = acl_get(path, item, self.numeric_owner)
```

Note that acl_get() *never* returned the ACL, it was always "fetch ACL and put it into the item".

This then results in:

```
{b'path': b'redacted', ... b'acl': None}
```

Suggested fix:

1. Add a test case for this; since we know exactly now how this looks the data is easily generated for that
1. Add to valid keys, and just ignore it.

**Comment by ThomasWaldmann** [comment]
Maybe don't add it to valid keys, it was never valid.

We could just obj.pop(b'acl', None) in valid_item() function, before the final check.

And add a comment above ITEM_KEYS constant that we shall never add a 'acl' key.

**Comment by enkore** [comment]
With the Item class in master we'll just add it to the in-line definition there:

```py
class Item(PropDict):
    VALID_KEYS = ITEM_KEYS | {'deleted', 'nlink', }
```

**Comment by cyril42e** [comment]
Fantastic, I tested your PR and I confirm that it does not find any problem on my repo anymore. I also ran a "borg extract --dry-run" with borg 1.0.8 on the first attic archive, last attic archive, first borg archive, and last borg archive to confirm that there were no other problem. Concerning the duplicate items of "borg debug-dump-archive-items", it seems to correspond at least for some of them to large duplicated files in the archive, so I guess it is normal. Thank you for solving this quickly!

**Comment by ThomasWaldmann** [comment]
@cyril42e thanks for testing the PR!

**Comment by enkore** [comment]
Fixed by #1841

**Comment by enkore** [comment]
(Also fixed in master now)
